app/src/main/java/com/example/sharingang/items/CachedItemRepository.kt

```diff
@@ -0,0 +1,51 @@
+package com.example.sharingang.items
+
+import androidx.lifecycle.LiveData
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+
+class CachedItemRepository @Inject constructor(
+    private val itemDao: ItemDao,
+    private val store: ItemStore
+) : ItemRepository {
+
+    override fun items(): LiveData<List<Item>> {
+        return itemDao.getAllItems()
+    }
+
+    override suspend fun refreshItems() {
+        // This is necessary, since you want to avoid doing this work on the main thread
+        withContext(Dispatchers.IO) {
+            val newItems = store.getAllItems()
+            itemDao.insert(newItems)
+        }
+    }
+
+    private suspend fun <T> thenRefresh(fn: suspend () -> T): T {
+        val ret = fn()
+        refreshItems()
+        return ret
+    }
+
+    override suspend fun addItem(item: Item): String? {
+        return thenRefresh { store.addItem(item) }
+    }
+
+    override suspend fun getItem(id: String): Item? {
+        return itemDao.getItem(id)
+    }
+
+    override suspend fun getAllItems(): List<Item> {
+        return thenRefresh { store.getAllItems() }
+    }
+
+    override suspend fun updateItem(item: Item): Boolean {
+        return thenRefresh { store.updateItem(item) }
+    }
+
+    override suspend fun deleteItem(id: String): Boolean {
```

Comment:
For the code duplication, what if we instead have one function editItem that does both actions, depending on a boolean value or enum value passed as parameter?